### PR TITLE
Move IsFileMatch to PathHelper

### PR DESCRIPTION
**Pull request by duncanp-sonar**
- renamed to IsServerFileMatch

Prep for #3997

### CODE REVIEW

File: src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs [hunk 1]
```diff
@@ -75,7 +74,7 @@ private static bool IsMatch(IFilterableIssue issue, SonarQubeIssue serverIssue)
                 return false;
             }
 
-            if (!IsFileMatch(issue, serverIssue))
+            if (!PathHelper.IsServerFileMatch(issue.FilePath, serverIssue.FilePath))
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Suppressions/SuppressedIssueMatcher.cs:77]: We need the same file-comparison logic when checking Roslyn suppressions, so I've moved the method to `Core.PathHelper`.

File: src/Core.UnitTests/Helpers/PathHelperTests.cs [hunk 2]
```diff
@@ -385,6 +385,34 @@ public void GetTempDirForTask_perVSInstanceTrueHasPath_AddsGuid()
...
+        [DataRow(@"aaa\partial\file.cs", @"d:\partial\file.cs", false)]
+        // Only matching the local path tail, so the same server path can match multiple local files
+        [DataRow(@"partial\file.cs", @"c:\aaa\partial\file.cs", true)]
+        [DataRow(@"partial\file.cs", @"c:\aaa\bbb\partial\file.cs", true)]
+        [DataRow(@"partial\file.cs", @"c:\aaa\bbb\ccc\partial\file.cs", true)]
+        public void IsFileMatch_ReturnsExpectedResult(string serverFilePath, string localFilePath, bool expected)
+        {
+            // Act and assert
+            PathHelper.IsServerFileMatch(localFilePath, serverFilePath).Should().Be(expected);
+        }
```
Comment (duncanp-sonar) [inline src/Core.UnitTests/Helpers/PathHelperTests.cs:415]: The test inputs are the same the corresponding `SuppressedIssueMatcher` test.
This duplication isn't ideal, but the alternatives are:
1) drop the corresponding `SuppressedIssueMatcher` test, - bad, as we'd have no coverage of this case.
2) refactor so we could inject an `IPathHelper` into the `SuppressedIssueMatcher`. This is the goal in the long run, but it's too much work to do at the moment.

@rita-sonarsource and I discussed the options yesterday and the current plan is:
* move the file-checking logic to `PathHelper` (i.e. this commit)
* change the Roslyn suppression checker to use the new `PathHelper` method (which will fix bug #3997)
* create a ticket to create an `IPathHelper` and refactor the existing code and tests.

In the meantime we'll live with the duplication in the test code.

File: src/Core/Helpers/PathHelper.cs [hunk 3]
```diff
@@ -146,13 +146,61 @@ public static bool IsPathRootedUnderRoot(string path, string rootDirectory)
...
+        /// <para>
+        /// Implementation: the server path is tail-matched against the local path
+        /// e.g. the server path "aaa\foo.txt" will match ALL of the following absolute paths:
+        /// * C:\aaa\foo.txt
+        /// * C:\bbb\aaa\foo.txt
+        /// * D:\bbb\foo.txt
+        /// It will not match e.g. C:\XXXaaa\foo.txt
+        /// </para>
+        /// </remarks>
+        public static bool IsServerFileMatch(string absoluteLocalFilePath, string relativeServerFilePath)
```
Comment (duncanp-sonar) [inline src/Core/Helpers/PathHelper.cs:167]: The implementation is the same, except I've changed the input parameters to be strings.
I added the doc comment.

File: src/Core/Helpers/PathHelper.cs [hunk 4]
```diff
@@ -146,13 +146,61 @@ public static bool IsPathRootedUnderRoot(string path, string rootDirectory)
...
+            var serverPath = relativeServerFilePath ?? string.Empty;
+
+            // A null/empty path means it's a module (project) level issue, and can only match
+            // another module-level issue.
+            if (localPath == string.Empty || serverPath == string.Empty)
+            {
+                return localPath == serverPath;
+            }
+
+            Debug.Assert(Path.IsPathRooted(localPath) && !localPath.Contains("/") && !localPath.Contains("..") && !localPath.Contains("\\.\\"),
```
Comment (duncanp-sonar) [inline src/Core/Helpers/PathHelper.cs:179]: I changed this assert to add two more cases (the last two).

File: src/Roslyn.Suppressions/Roslyn.Suppressions.UnitTests/SuppressionCheckerTests.cs [hunk 5]
```diff
@@ -104,7 +106,7 @@ public void IsSuppressed_HasIssues_NoMatches_ReturnsFalse()
             
             var suppressedIssues = new SuppressedIssue[]
             {
-                CreateIssue("S111", "c:\\wrongFile1.txt")
+                CreateIssue("S111", "wrongFile1.txt")
```
Comment (duncanp-sonar) [inline src/Roslyn.Suppressions/Roslyn.Suppressions.UnitTests/SuppressionCheckerTests.cs:109]: Fixing the test input; server paths are relative.

File: src/Roslyn.Suppressions/Roslyn.Suppressions.UnitTests/SuppressionCheckerTests.cs [hunk 6]
```diff
@@ -150,7 +152,7 @@ public void IsMatch_FilesAreDifferent_ReturnsFalse()
             // Sanity check the diagnostic location is on the expected line
             diag.Location.GetLineSpan().StartLinePosition.Line.Should().Be(0, "Test setup error");
 
-            var issue = CreateIssue("S999", "c:\\issueFile.cs", 1, "hash");
+            var issue = CreateIssue("S999", "issueFile.cs", 1, "hash");
```
Comment (duncanp-sonar) [inline src/Roslyn.Suppressions/Roslyn.Suppressions.UnitTests/SuppressionCheckerTests.cs:155]: Ditto - fixing the test input.

File: src/Roslyn.Suppressions/Roslyn.Suppressions/SuppressionChecker.cs [hunk 7]
```diff
@@ -117,8 +118,7 @@ public bool IsSuppressed(Diagnostic reportedDiagnostic, string settingsKey)
         }
 
         private static bool IsSameFile(Diagnostic diagnostic, SuppressedIssue suppressedIssue) =>
-            diagnostic.Location.SourceTree?.FilePath.EndsWith(suppressedIssue.FilePath, StringComparison.OrdinalIgnoreCase)
-            ?? false;
+            PathHelper.IsServerFileMatch(diagnostic.Location.SourceTree?.FilePath, suppressedIssue.FilePath);
```
Comment (duncanp-sonar) [inline src/Roslyn.Suppressions/Roslyn.Suppressions/SuppressionChecker.cs:121]: Doh! This is the fix for the Roslyn suppression. I meant to exclude it from this PR...

File: src/Roslyn.Suppressions/Roslyn.Suppressions.UnitTests/SuppressionCheckerTests.cs [hunk 8]
```diff
@@ -80,7 +80,7 @@ public void IsSuppressed_NullSettings_ReturnsFalse()
             var actual = testSubject.IsSuppressed(ValidInSourceDiagnostic, "settingsKey1");
 
             actual.Should().Be(false);
-            CheckGetSettingsCalled(cache, "settingsKey1");        
+            CheckGetSettingsCalled(cache, "settingsKey1");
```
Comment (duncanp-sonar) [inline src/Roslyn.Suppressions/Roslyn.Suppressions.UnitTests/SuppressionCheckerTests.cs:83]: Unintended whitespace changes by VS

**Review by bigfluffycookie — APPROVED** [review]

**Pull request merged**